src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs

```diff
@@ -651,6 +652,27 @@ private static bool MatchSwitch(string switchKey, string match, string smallestU
                     && match.StartsWith(switchKey, StringComparison.OrdinalIgnoreCase));
         }
 
+        private static bool IsStaSupported()
+        {
+#if UNIX
+            return false;
+#else
+            // On some variants of Windows, STA is not supported and pwsh will fail to start.
+            // The exception to determine if STA is supported only occurs on thread Start().
+            Thread thread = new Thread(() => { });
+            thread.SetApartmentState(ApartmentState.STA);
+            try
+            {
+                thread.Start();
+                return true;
+            }
+            catch
+            {
+                return false;
+            }
```

Comment:
Is this on the startup path and for all Windows?

It seems like an expensive test to do on every startup on all Windows platforms.

Could we either:

- Find a better test
- Narrow down the circumstances when we perform it
- Or give a justification in the code about why we have to do this test and open an issue to look into it again later

Reply:
It's in the startup path if `-MTA` isn't specified.  Since STA is default, we have to check before we try to use it.  I tried to find a better way to determine if the particular build of Windows supports STA, but couldn't find anything.  I believe the comment justifies it unless you feel more explanation is needed?

Reply:
My only feeling on the comment is that it should say something like:

```
Launching a new thread and checking for an exception is currently the only way to check if STA is supported.
```

I also think we should open an issue with dotnet/runtime to ask about this, since it's not something we should have to do.

Which platforms is it not supported on? Is there a way to avoid this test in most cases?

Reply:
It's not really the problem of dotnet as it's a COM thing.  As for the platforms, there are public ones like Nanoserver, but also not-public ones that are so easily determined.  I tried to look for another way to check, but couldn't find it.

I thought line 661 says what you want said?

Reply:
> It's not really the problem of dotnet as it's a COM thing.

That's fair enough -- but surely there should be an easier way for us to query this

> I thought line 661 says what you want said?

I would have phrased it differently in the interests of later readers, but it's probably fine as it is

Reply:
We could directly test platforms we want return STA false for
https://github.com/dotnet/runtime/blob/79ae74f5ca5c8a6fe3a48935e85bd7374959c570/src/libraries/Common/tests/TestUtilities/System/PlatformDetection.Windows.cs#L20-L37

Reply:
I just saw today that we already do a check just like this in the code today:

https://github.com/PowerShell/PowerShell/blob/a8a7e045d8ed2335cc4092bf467f2876a61c733e/src/System.Management.Automation/engine/hostifaces/LocalPipeline.cs#L195-L198

Reply:
The problem is that IsNanoServer and IsIoT is checking specific registry values.  In this case, the sku that has the problem is not IoT nor NanoServer and I'd hate to keep track of all the various skus (think Hololens, XBox, Surface Hub, etc...).  I think an alternative approach might be to see if `ProductName` doesn't start with `Windows` and assume STA isn't supported.  This should cover IoT, Nano, and new variants.